### PR TITLE
Add message output estimation

### LINKED ISSUE #845: Add output message tx dependency estimation
Currently we support automatic setup of external contracts and output variables. The same can be implemented for output messages.

### PR DESCRIPTION
**Pull request by MujkicA**
Closes #845

Extends `estimate_tx_dependancies` to also add output messages if they're missing.

### CODE REVIEW

File: packages/fuels/tests/contracts.rs [hunk 1]
```diff
@@ -703,11 +741,15 @@ async fn test_output_variable_estimation_multicall() -> Result<()> {
     let amount = 1000;
 
     let mut multi_call_handler = MultiContractCallHandler::new(wallets[0].clone());
-    (0..3).for_each(|_| {
+    (0..2).for_each(|_| {
```
Comment (digorithm) [inline packages/fuels/tests/contracts.rs:744]: Why this change here? 👀

**Review by digorithm — APPROVED** [review]
LGTM -- just left a few clarification questions!

File: packages/fuels/tests/contracts.rs (hunk 1 above)
Comment (MujkicA) [reply packages/fuels/tests/contracts.rs:744]: Actually, there is no need for it

File: packages/fuels-programs/src/contract.rs [hunk 2]
```diff
@@ -886,6 +898,15 @@ impl MultiContractCallHandler {
                         .for_each(|call| call.append_variable_outputs(1));
                 }
 
+                Err(Error::RevertTransactionError { receipts, .. })
+                    if ContractCall::is_missing_message_output(&receipts) =>
+                {
+                    self.contract_calls
+                        .iter_mut()
+                        .take(1)
```
[the remark this answers is not in the document]
Comment (MujkicA) [reply packages/fuels-programs/src/contract.rs:906]: I've simplified this a little and added a comment. 
The gist is that for multi calls, it doesn't matter to which call we append the output. They all get merged into a single tx.

File: packages/fuels-programs/src/contract.rs [hunk 3]
```diff
@@ -858,10 +870,15 @@ impl MultiContractCallHandler {
                 Err(Error::RevertTransactionError { receipts, .. })
                     if ContractCall::is_missing_output_variables(&receipts) =>
                 {
-                    self.contract_calls
-                        .iter_mut()
-                        .take(1)
-                        .for_each(|call| call.append_variable_outputs(1));
+                    // It doesn't matter where the missing output is added,
+                    // all calls will be merged to a single script tx
+                    self.contract_calls[0].append_variable_outputs(1);
```
Comment (segfault-magnet) [inline packages/fuels-programs/src/contract.rs:875]: What makes you sure that `self.contract_calls.len() > 0` will always be true at this point?

If there is a reason can you add it to a comment? 

Otherwise you can do: 
```rust
                    if let Some(first_contract_call) = self.contract_calls.first_mut() {
                        first_contract_call.append_message_outputs(1);
                    }
```

File: packages/fuels-programs/src/contract.rs [hunk 4]
```diff
@@ -858,10 +870,15 @@ impl MultiContractCallHandler {
...
+                }
+
+                Err(Error::RevertTransactionError { receipts, .. })
+                    if ContractCall::is_missing_message_output(&receipts) =>
+                {
+                    self.contract_calls[0].append_message_outputs(1);
                 }
 
                 Err(Error::RevertTransactionError { ref receipts, .. }) => {
```
Comment (segfault-magnet) [inline packages/fuels-programs/src/contract.rs:884]: A simpler way to write this:
```rust
                    if let Some(first_contract_call) = self.contract_calls.first_mut() {
                        call.append_external_contracts(contract_id.clone());
                    }
```

**Review by segfault-magnet — COMMENTED** [review]
Reviewed alongside @hal3e. We've put our bigger impacting suggestions in a separate branch, you can compare the differences [here](https://github.com/FuelLabs/fuels-rs/compare/mujkica/add_message_output_estimation...mujkica/add_message_output_estimation_pr_suggestions).

Reasons for the suggestions:
1. If a revert happened not because of a missing variable or message output then the code assumes that it must have been a missing contract input. This isn't true.
2. The message inside `expect_err` in the abovementioned case will never be propagated. The result in question will always be an `Err` at that place in the code so `expect_err` will never panic. If the original intent was for the message to be propagated then our changes offer one way of achieving that.
3. The current code assumes there can only be one revert reason (since only one match branch can be selected). Although this is probably true atm, it costs nothing to write it in a manner that can support multiple reverts.

We haven't discussed the MultiCallHandler's estimations. Perhaps they might benefit from the same suggestions.

File: packages/fuels-programs/src/contract.rs (hunk 3 above)
Comment (MujkicA) [reply packages/fuels-programs/src/contract.rs:879]: Using simulate or call requires that the underlying tx is built via `build_tx()` which fails if there are no contract calls.

File: packages/fuels-programs/src/contract.rs (hunk 4 above)
Comment (MujkicA) [reply packages/fuels-programs/src/contract.rs:888]: This seems like the same thing as mentioned above

File: packages/fuels-programs/src/contract.rs (hunk 3 above)
Comment (iqdecay) [reply packages/fuels-programs/src/contract.rs:879]: This trick should be documented

Comment (MujkicA) [reply packages/fuels-programs/src/contract.rs:879]: I've applied the changes to both single and multi calls. I did change the naming a little and opted of `iter_mut().take(1)` instead of the above.

Comment (Salka1988) [reply packages/fuels-programs/src/contract.rs]: Papi's got tricks up his sleeves.

**Review by iqdecay — APPROVED** [review]
LGTM

**Review by Salka1988 — APPROVED** [review]
:+1:
